flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/sink/compactor/operator/CompactorOperator.java

```diff
@@ -0,0 +1,404 @@
+        Iterator<Entry<Long, List<CompactorRequest>>> iter = canSubmit.entrySet().iterator();
+        while (iter.hasNext()) {
+            Entry<Long, List<CompactorRequest>> requestEntry = iter.next();
+            for (CompactorRequest req : requestEntry.getValue()) {
+                CompletableFuture<Iterable<FileSinkCommittable>> resultFuture =
+                        new CompletableFuture<>();
+                compactingRequests.add(new Tuple2<>(req, resultFuture));
+                compactService.submit(
+                        () -> {
+                            try {
+                                Iterable<FileSinkCommittable> result = compact(req);
+                                resultFuture.complete(result);
+                            } catch (Exception e) {
+                                resultFuture.completeExceptionally(e);
+                            }
+                        });
+            }
+            iter.remove();
+        }
+    }
+
+    private Iterable<FileSinkCommittable> compact(CompactorRequest request) throws Exception {
+        List<FileSinkCommittable> results = new ArrayList<>(request.getCommittableToPassthrough());
+
+        List<Path> compactingFiles = getCompactingPath(request, results);
+        if (compactingFiles.isEmpty()) {
+            return results;
+        }
+
+        Path targetPath = assembleCompactedFilePath(compactingFiles.get(0));
+        CompactingFileWriter compactingFileWriter =
+                bucketWriter.openNewCompactingFile(
+                        fileCompactor.getWriterType(),
+                        request.getBucketId(),
+                        targetPath,
+                        System.currentTimeMillis());
+        fileCompactor.compact(compactingFiles, compactingFileWriter);
+        PendingFileRecoverable compactedPendingFile = compactingFileWriter.closeForCommit();
+
+        FileSinkCommittable compacted =
+                new FileSinkCommittable(request.getBucketId(), compactedPendingFile);
+        results.add(0, compacted);
+        for (Path f : compactingFiles) {
+            // cleanup compacted files
+            results.add(new FileSinkCommittable(request.getBucketId(), f));
+        }
+
+        return results;
+    }
+
+    // results: side output pass through committable
+    private List<Path> getCompactingPath(
+            CompactorRequest request, List<FileSinkCommittable> results) throws IOException {
+        List<FileSinkCommittable> compactingCommittable = request.getCommittableToCompact();
+        List<Path> compactingFiles = new ArrayList<>();
+
+        for (FileSinkCommittable committable : compactingCommittable) {
+            PendingFileRecoverable pendingFile = committable.getPendingFile();
+            if (pendingFile == null
+                    || pendingFile.getPath() == null
+                    || !pendingFile.getPath().getName().startsWith(".")) {
```

Comment:
Why don't we pass-through these pending files in coordinator? Since it might cause the `CompactStrategy` not fulfilled if we consider these files here.